go-controller/pkg/cluster/ha_master.go

```diff
@@ -0,0 +1,321 @@
+		// After the leader election, the leader will promote the OVN Dbs
+		// to become active.
+		err := hacluster.DemoteOVNDbs(invalidIPAddres, "6641", "6642")
+		if err != nil {
+			// If we are not able to communicate to the OVN ovsdb-servers,
+			// then it is better to return than continue.
+			// cmd/ovnkube.go will panic if this function returns error.
+			return err
+		}
+	}
+
+	// Set up leader election process first
+	rl, err := resourcelock.New(
+		resourcelock.ConfigMapsResourceLock,
+		config.Kubernetes.OVNConfigNamespace,
+		haLeaderLockName,
+		hacluster.kubeClient.CoreV1(),
+		nil,
+		resourcelock.ResourceLockConfig{
+			Identity:      hacluster.nodeName,
+			EventRecorder: nil,
+		})
+
+	if err != nil {
+		return err
+	}
+
+	HAClusterNewLeader := func(identity string) {
+		logrus.Infof("The new leader is : " + identity)
+		wasLeader := hacluster.isLeader
+		if hacluster.nodeName == identity {
+			err = hacluster.ConfigureAsActive(identity)
+			if err != nil {
+				logrus.Errorf(err.Error())
+				panic(err.Error())
+			}
+			hacluster.isLeader = true
+		} else {
+			if wasLeader {
+				//This node was leader and it lost the election.
+				// Whenever the node transitions from leader to follower,
+				// we need to handle the transition properly like clearing
+				// the cache. It is better to exit for now.
+				// kube will restart and this will become a follower.
+				logrus.Infof("[%s] was the leader and it lost the election. Exiting", hacluster.nodeName)
+				os.Exit(1)
+			}
+
+			err = hacluster.ConfigureAsStandby(identity)
+			if err != nil {
+				logrus.Errorf(err.Error())
+				panic(err.Error())
+			}
+
+		}
+	}
+
+	lec := leaderelection.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: leaseDuration,
+		RenewDeadline: renewDeadline,
+		RetryPeriod:   retryPeriod,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {},
+			OnStoppedLeading: func() {
+				//This node was leader and it lost the election.
+				// Whenever the node transitions from leader to follower,
+				// we need to handle the transition properly like clearing
+				// the cache. It is better to exit for now.
+				// kube will restart and this will become a follower.
+				logrus.Infof("[%s] stopped Leading. Exiting", hacluster.nodeName)
+				os.Exit(1)
+			},
+			OnNewLeader: HAClusterNewLeader,
+		},
+	}
+
+	go leaderelection.RunOrDie(context.Background(), lec)
+
+	return nil
+}
+
+// ConfigureAsActive configures the node as active.
+func (hacluster *OvnHAMasterController) ConfigureAsActive(masterNodeName string) error {
+	if hacluster.manageDBServers {
+		// Find the endpoint for the service
+		ep, err := hacluster.cluster.Kube.GetEndpoint(config.Kubernetes.OVNConfigNamespace, ovnkubeDbEp)
+		createEp := true
+		if err == nil {
+			if ep.Name == ovnkubeDbEp {
+				createEp = false
+			}
+		}
+
+		const (
+			nbPort int32 = 6641
+			sbPort int32 = 6642
+		)
+
+		ovndbEp := v1.Endpoints{
```

Comment:
This is a bit subtle - it will delete any of the annotations and labels we may have placed on the object. Instead, what you should do is
1. Get() the endpoint by name
2. If not found, create
3. If found, deep-copy and make your changes
4. Update()

Reply:
Ack.